File: core/src/nl/t64/game/rpg/components/party/InventoryItem.java
package nl.t64.game.rpg.components.party;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;


@NoArgsConstructor
@Setter
public class InventoryItem {

    @Getter
    String id;
    String name;
    int sort;
    @Getter
    List<String> description;
    @Getter
    int amount;
    @Getter
    InventoryGroup group;
    @JsonProperty("is_two_handed")
    boolean isTwoHanded;
    SkillItemId skill;
    private int price;
    private int weight;
    @JsonProperty("min_intelligence")
    private int minIntelligence;
    @JsonProperty("min_willpower")
    private int minWillpower;
    @JsonProperty("min_dexterity")
    private int minDexterity;
    @JsonProperty("min_strength")
    private int minStrength;
    private int movepoints;
    @JsonProperty("base_hit")
    private int baseHit;
    private int damage;
    private int protection;
    private int defense;
    @JsonProperty("spell_battery")
    private int spellBattery;
    private int intelligence;
    private int willpower;
    private int dexterity;
    private int strength;
    private int agility;
    private int endurance;
    private int alchemist;
    private int diplomat;
    private int healer;
    private int loremaster;
    private int mechanic;
    private int ranger;
    private int stealth;
    private int thief;
    private int troubadour;
    private int warrior;
    private int wizard;
    @JsonProperty("cheat_death")
    private boolean cheatDeath;
    @JsonProperty("quick_switch")
    private boolean quickSwitch;
    @JsonProperty("spell_boost")
    private boolean spellBoost;

    InventoryItem(InventoryItem item, int amount) {
        this.id = item.id;
        this.name = item.name;
        this.sort = item.sort;
        this.description = item.description;
        this.amount = amount;
        this.group = item.group;
        this.skill = item.skill;
        this.isTwoHanded = item.isTwoHanded;
        this.price = item.price;
        this.weight = item.weight;
        this.minIntelligence = item.minIntelligence;
        this.minWillpower = item.minWillpower;
        this.minDexterity = item.minDexterity;
        this.minStrength = item.minStrength;
        this.movepoints = item.movepoints;
        this.baseHit = item.baseHit;
        this.damage = item.damage;
        this.protection = item.protection;
        this.defense = item.defense;
        this.spellBattery = item.spellBattery;
        this.intelligence = item.intelligence;
        this.willpower = item.willpower;
        this.dexterity = item.dexterity;
        this.strength = item.strength;
        this.agility = item.agility;
        this.endurance = item.endurance;
        this.alchemist = item.alchemist;
        this.diplomat = item.diplomat;
        this.healer = item.healer;
        this.loremaster = item.loremaster;
        this.mechanic = item.mechanic;
        this.ranger = item.ranger;
        this.stealth = item.stealth;
        this.thief = item.thief;
        this.troubadour = item.troubadour;
        this.warrior = item.warrior;
        this.wizard = item.wizard;
        this.cheatDeath = item.cheatDeath;
        this.quickSwitch = item.quickSwitch;
        this.spellBoost = item.spellBoost;
    }

    public static InventoryItem copyOf(InventoryItem inventoryItem, int amount) {
        return new InventoryItem(inventoryItem, amount);
    }

    Object getAttributeOfMinimal(InventoryMinimal minimal) {
        return switch (minimal) {
            case SKILL -> Objects.requireNonNullElse(skill, 0);
            case MIN_INTELLIGENCE -> minIntelligence;
            case MIN_WILLPOWER -> minWillpower;
            case MIN_DEXTERITY -> minDexterity;
            case MIN_STRENGTH -> minStrength;
        };
    }

    int getMinimalAttributeOfStatItemId(StatItemId statItemId) {
        return switch (statItemId) {
            case INTELLIGENCE -> minIntelligence;
            case WILLPOWER -> minWillpower;
            case DEXTERITY -> minDexterity;
            case STRENGTH -> minStrength;
            case AGILITY, ENDURANCE, STAMINA -> 0;
        };
    }

    int getAttributeOfStatItemId(StatItemId statItemId) {
        return switch (statItemId) {
            case INTELLIGENCE -> intelligence;
            case WILLPOWER -> willpower;
            case DEXTERITY -> dexterity;
            case STRENGTH -> strength;
            case AGILITY -> agility;
            case ENDURANCE -> endurance;
            case STAMINA -> 0;
        };
    }

    int getAttributeOfSkillItemId(SkillItemId skillItemId) {
        return switch (skillItemId) {
            case ALCHEMIST -> alchemist;
            case DIPLOMAT -> diplomat;
            case HEALER -> healer;
            case LOREMASTER -> loremaster;
            case MECHANIC -> mechanic;
            case STEALTH -> stealth;
            case RANGER -> ranger;
            case THIEF -> thief;
            case TROUBADOUR -> troubadour;
            case WARRIOR -> warrior;
            case WIZARD -> wizard;
            case MERCHANT, HAFTED, MISSILE, POLE, SHIELD, SWORD, THROWN -> 0;
        };
    }

    int getAttributeOfCalcAttributeId(CalcAttributeId calcAttributeId) {
        return switch (calcAttributeId) {
            case WEIGHT -> weight;
            case MOVEPOINTS -> movepoints;
            case BASE_HIT -> baseHit;
            case DAMAGE -> damage;
            case PROTECTION -> protection;
            case DEFENSE -> defense;
            case SPELL_BATTERY -> spellBattery;
        };
    }

    public Set<InventoryMinimal> getMinimalsOtherItemHasAndYouDont(InventoryItem otherItem) {
        if (hasOnlyOneMinimal() && otherItem.hasOnlyOneMinimal()) {
            return Set.of();
        } else {
            return Arrays.stream(InventoryMinimal.values())
                         .filter(minimal -> getAttributeOfMinimal(minimal) instanceof Integer)
                         .filter(minimal -> (int) getAttributeOfMinimal(minimal) == 0)
                         .filter(minimal -> (int) otherItem.getAttributeOfMinimal(minimal) > 0)
                         .collect(Collectors.toUnmodifiableSet());
        }
    }

    private boolean hasOnlyOneMinimal() {
        return Arrays.stream(InventoryMinimal.values())
                     .filter(minimal -> getAttributeOfMinimal(minimal) instanceof Integer)
                     .filter(minimal -> (int) getAttributeOfMinimal(minimal) > 0)
                     .count() == 1L;
    }

    public Set<CalcAttributeId> getCalcsOtherItemHasAndYouDont(InventoryItem otherItem) {
        return Arrays.stream(CalcAttributeId.values())
                     .filter(calcAttributeId -> getAttributeOfCalcAttributeId(calcAttributeId) == 0)
                     .filter(calcAttributeId -> otherItem.getAttributeOfCalcAttributeId(calcAttributeId) > 0)
                     .collect(Collectors.toUnmodifiableSet());
    }

    public Set<StatItemId> getStatsOtherItemHasAndYouDont(InventoryItem otherItem) {
        return Arrays.stream(StatItemId.values())
                     .filter(statItemId -> getAttributeOfStatItemId(statItemId) == 0)
                     .filter(statItemId -> otherItem.getAttributeOfStatItemId(statItemId) > 0)
                     .collect(Collectors.toUnmodifiableSet());
    }

    public Set<SkillItemId> getSkillsOtherItemHasAndYouDont(InventoryItem otherItem) {
        return Arrays.stream(SkillItemId.values())
                     .filter(skillItemId -> getAttributeOfSkillItemId(skillItemId) == 0)
                     .filter(skillItemId -> otherItem.getAttributeOfSkillItemId(skillItemId) > 0)
                     .collect(Collectors.toUnmodifiableSet());
    }

    public boolean hasSameIdAs(InventoryItem candidateItem) {
        return hasSameIdAs(candidateItem.id);
    }

    boolean hasSameIdAs(String candidateId) {
        return id.equalsIgnoreCase(candidateId);
    }

    public boolean isStackable() {
        return group.isStackable();
    }

    public int getBuyPriceTotal(int totalMerchant) {
        return getBuyPricePiece(totalMerchant) * amount;
    }

    public int getBuyPricePiece(int totalMerchant) {
        int salePrice = Math.round(price - ((price / 100f) * totalMerchant));
        if (salePrice == 0) {
            salePrice = 1;
        }
        return salePrice;
    }

    public int getSellValueTotal(int totalMerchant) {
        return getSellValuePiece(totalMerchant) * amount;
    }

    public int getSellValuePiece(int totalMerchant) {
        int value = (int) Math.floor(price / 3f);
        return Math.round(value + ((value / 300f) * totalMerchant));
    }

    void increaseAmountWith(InventoryItem sourceItem) {
        increaseAmountWith(sourceItem.amount);
    }

    void increaseAmountWith(int amount) {
        this.amount += amount;
    }

    void decreaseAmountWith(int amount) {
        this.amount -= amount;
        if (this.amount < 1) {
            throw new IllegalStateException("Amount cannot be below 1.");
        }
    }

    String createMessageFailToDequip(InventoryItem dependantItem) {
        return String.format("""
                                     Cannot unequip the %s.
                                     The %s depends on it.""", name, dependantItem.name);
    }

    String createMessageFailToEquipTwoHanded(InventoryItem otherItem) {
        return String.format("""
                                     Cannot equip the %s.
                                     First unequip the %s.""", name, otherItem.name);
    }

}

File: core/src/nl/t64/game/rpg/components/party/inventory/InventoryItem.kt
package nl.t64.game.rpg.components.party.inventory

import com.fasterxml.jackson.annotation.JsonProperty
import nl.t64.game.rpg.components.party.CalcAttributeId
import nl.t64.game.rpg.components.party.skills.SkillItemId
import nl.t64.game.rpg.components.party.stats.StatItemId
import kotlin.math.floor
import kotlin.math.roundToInt


class InventoryItem(
    val name: String = "",
    val sort: Int = 0,
    val description: List<String> = emptyList(),
    val group: InventoryGroup = InventoryGroup.EMPTY,
    @JsonProperty("is_two_handed")
    val isTwoHanded: Boolean = false,
    val skill: SkillItemId? = null,
    private val price: Int = 0,
    private val weight: Int = 0,
    @JsonProperty("min_intelligence")
    private val minIntelligence: Int = 0,
    @JsonProperty("min_willpower")
    private val minWillpower: Int = 0,
    @JsonProperty("min_dexterity")
    private val minDexterity: Int = 0,
    @JsonProperty("min_strength")
    private val minStrength: Int = 0,
    private val movepoints: Int = 0,
    @JsonProperty("base_hit")
    private val baseHit: Int = 0,
    private val damage: Int = 0,
    private val protection: Int = 0,
    private val defense: Int = 0,
    @JsonProperty("spell_battery")
    private val spellBattery: Int = 0,
    private val intelligence: Int = 0,
    private val willpower: Int = 0,
    private val dexterity: Int = 0,
    private val strength: Int = 0,
    private val agility: Int = 0,
    private val endurance: Int = 0,
    private val alchemist: Int = 0,
    private val diplomat: Int = 0,
    private val healer: Int = 0,
    private val loremaster: Int = 0,
    private val mechanic: Int = 0,
    private val ranger: Int = 0,
    private val stealth: Int = 0,
    private val thief: Int = 0,
    private val troubadour: Int = 0,
    private val warrior: Int = 0,
    private val wizard: Int = 0,
    @JsonProperty("cheat_death")
    private val cheatDeath: Boolean = false,
    @JsonProperty("quick_switch")
    private val quickSwitch: Boolean = false,
    @JsonProperty("spell_boost")
    private val spellBoost: Boolean = false
) {

    lateinit var id: String
    var amount: Int = 0
    val isStackable: Boolean get() = group.isStackable()

    fun createCopy(amount: Int): InventoryItem {
        val itemCopy = InventoryItem(
            name, sort, description, group, isTwoHanded, skill, price, weight,
            minIntelligence, minWillpower, minDexterity, minStrength,
            movepoints, baseHit, damage, protection, defense, spellBattery,
            intelligence, willpower, dexterity, strength, agility, endurance,
            alchemist, diplomat, healer, loremaster, mechanic, ranger, stealth, thief, troubadour, warrior, wizard,
            cheatDeath, quickSwitch, spellBoost)
        itemCopy.id = id
        itemCopy.amount = amount
        return itemCopy
    }

    fun getAttributeOfMinimal(minimal: InventoryMinimal): Any {
        return when (minimal) {
            InventoryMinimal.SKILL -> skill ?: 0
            InventoryMinimal.MIN_INTELLIGENCE -> minIntelligence
            InventoryMinimal.MIN_WILLPOWER -> minWillpower
            InventoryMinimal.MIN_DEXTERITY -> minDexterity
            InventoryMinimal.MIN_STRENGTH -> minStrength
        }
    }

    fun getMinimalAttributeOfStatItemId(statItemId: StatItemId): Int {
        return when (statItemId) {
            StatItemId.INTELLIGENCE -> minIntelligence
            StatItemId.WILLPOWER -> minWillpower
            StatItemId.DEXTERITY -> minDexterity
            StatItemId.STRENGTH -> minStrength
            StatItemId.AGILITY,
            StatItemId.ENDURANCE,
            StatItemId.STAMINA -> 0
        }
    }

    fun getAttributeOfStatItemId(statItemId: StatItemId): Int {
        return when (statItemId) {
            StatItemId.INTELLIGENCE -> intelligence
            StatItemId.WILLPOWER -> willpower
            StatItemId.DEXTERITY -> dexterity
            StatItemId.STRENGTH -> strength
            StatItemId.AGILITY -> agility
            StatItemId.ENDURANCE -> endurance
            StatItemId.STAMINA -> 0
        }
    }

    fun getAttributeOfSkillItemId(skillItemId: SkillItemId): Int {
        return when (skillItemId) {
            SkillItemId.ALCHEMIST -> alchemist
            SkillItemId.DIPLOMAT -> diplomat
            SkillItemId.HEALER -> healer
            SkillItemId.LOREMASTER -> loremaster
            SkillItemId.MECHANIC -> mechanic
            SkillItemId.STEALTH -> stealth
            SkillItemId.RANGER -> ranger
            SkillItemId.THIEF -> thief
            SkillItemId.TROUBADOUR -> troubadour
            SkillItemId.WARRIOR -> warrior
            SkillItemId.WIZARD -> wizard
            SkillItemId.MERCHANT,
            SkillItemId.HAFTED,
            SkillItemId.MISSILE,
            SkillItemId.POLE,
            SkillItemId.SHIELD,
            SkillItemId.SWORD,
            SkillItemId.THROWN -> 0
            SkillItemId.BITE -> 0
        }
    }

    fun getAttributeOfCalcAttributeId(calcAttributeId: CalcAttributeId): Int {
        return when (calcAttributeId) {
            CalcAttributeId.WEIGHT -> weight
            CalcAttributeId.MOVEPOINTS -> movepoints
            CalcAttributeId.BASE_HIT -> baseHit
            CalcAttributeId.DAMAGE -> damage
            CalcAttributeId.PROTECTION -> protection
            CalcAttributeId.DEFENSE -> defense
            CalcAttributeId.SPELL_BATTERY -> spellBattery
        }
    }

    fun getMinimalsOtherItemHasAndYouDont(otherItem: InventoryItem): Set<InventoryMinimal> {
        return when {
            hasOnlyOneMinimal() && otherItem.hasOnlyOneMinimal() -> emptySet()
            else -> InventoryMinimal.values()
                .filter { getAttributeOfMinimal(it) is Int }
                .filter { getAttributeOfMinimal(it) as Int == 0 }
                .filter { otherItem.getAttributeOfMinimal(it) as Int > 0 }
                .toSet()
        }
    }

    private fun hasOnlyOneMinimal(): Boolean {
        return InventoryMinimal.values()
            .filter { getAttributeOfMinimal(it) is Int }
            .filter { getAttributeOfMinimal(it) as Int > 0 }
            .count() == 1
    }

    fun getCalcsOtherItemHasAndYouDont(otherItem: InventoryItem): Set<CalcAttributeId> {
        return CalcAttributeId.values()
            .filter { getAttributeOfCalcAttributeId(it) == 0 }
            .filter { otherItem.getAttributeOfCalcAttributeId(it) > 0 }
            .toSet()
    }

    fun getStatsOtherItemHasAndYouDont(otherItem: InventoryItem): Set<StatItemId> {
        return StatItemId.values()
            .filter { getAttributeOfStatItemId(it) == 0 }
            .filter { otherItem.getAttributeOfStatItemId(it) > 0 }
            .toSet()
    }

    fun getSkillsOtherItemHasAndYouDont(otherItem: InventoryItem): Set<SkillItemId> {
        return SkillItemId.values()
            .filter { getAttributeOfSkillItemId(it) == 0 }
            .filter { otherItem.getAttributeOfSkillItemId(it) > 0 }
            .toSet()
    }

    fun hasSameIdAs(candidateItem: InventoryItem): Boolean {
        return hasSameIdAs(candidateItem.id)
    }

    fun hasSameIdAs(candidateId: String?): Boolean {
        return id.equals(candidateId, true)
    }

    fun getBuyPriceTotal(totalMerchant: Int): Int {
        return getBuyPricePiece(totalMerchant) * amount
    }

    fun getBuyPricePiece(totalMerchant: Int): Int {
        return (price - ((price / 100f) * totalMerchant)).roundToInt().run {
            takeUnless { this == 0 } ?: 1
        }
    }

    fun getSellValueTotal(totalMerchant: Int): Int {
        return getSellValuePiece(totalMerchant) * amount
    }

    fun getSellValuePiece(totalMerchant: Int): Int {
        val value = floor(price / 3f).toInt()
        return (value + ((value / 300f) * totalMerchant)).roundToInt()
    }

    fun increaseAmountWith(sourceItem: InventoryItem) {
        increaseAmountWith(sourceItem.amount)
    }

    fun increaseAmountWith(amount: Int) {
        this.amount += amount
    }

    fun decreaseAmountWith(amount: Int) {
        this.amount -= amount
        check(this.amount >= 1) { "Amount cannot be below 1." }
    }

    fun createMessageFailToDequip(dependantItem: InventoryItem): String {
        return """
            Cannot unequip the $name.
            The ${dependantItem.name} depends on it.""".trimIndent()
    }

    fun createMessageFailToEquipTwoHanded(otherItem: InventoryItem): String {
        return """
            Cannot equip the $name.
            First unequip the ${otherItem.name}.""".trimIndent()
    }

}
